### ISSUE TITLE
Severe VRAM regression / OOM with Krea 2 training after PyTorch 2.13 update

### ISSUE THREAD

**Issue by Tholue** [ostris/ai-toolkit#990]
## This is for bugs only

Did you already ask [in the discord](https://discord.gg/VXmU2f5WEU)?

No

You verified that this is a bug and not a feature request or question by asking [in the discord](https://discord.gg/VXmU2f5WEU)?

Yes

## Describe the bug
# Severe VRAM regression / OOM after updating to PyTorch 2.13

## Description

After updating ai-toolkit and its dependencies, including the major PyTorch update from 2.12.1 to 2.13.0, Krea 2 LoRA training now runs out of VRAM on an RTX 3090 24 GB.

The same training configurations and datasets worked before the update at considerably higher resolutions.

After the update:

* Training frequently OOMs during the first few training steps.
* Lowering the dataset resolution does not reliably prevent the OOM.
* Enabling or disabling `low_vram` does not solve it.
* Restarting the computer does not solve it.
* In some runs, memory allocation warnings already occur while the dataset or image data is being processed.
* The GPU has almost no free VRAM left when the first training step begins.

This appears to be a significant VRAM usage or memory-management regression introduced by the new dependency stack.

## Environment

* OS: CachyOS / Arch Linux
* Python: 3.12.13
* GPU: NVIDIA RTX 3090 24 GB
* CUDA: 13.0
* ai-toolkit: current version after update on August 2, 2026

Dependency versions after the update:

```text
torch==2.13.0
torchvision==0.28.0
cuda-toolkit==13.0.3.0
setuptools==83.0.0
```

Previously working versions:

```text
torch==2.12.1+cu130
torchvision==0.27.1+cu130
cuda-toolkit==13.0.2
setuptools==69.5.1
```

## Model and training configuration

* Architecture: `krea2`
* Model: Krea 2 Raw
* Transformer quantization: `qfloat8`
* Text encoder quantization: `qfloat8`
* Training dtype: `bf16`
* LoRA rank: 32
* Batch size: 1
* Gradient accumulation: 1
* Gradient checkpointing: enabled
* Optimizer: `adamw8bit`
* Text embeddings cached
* Latents cached to disk
* Text encoder training disabled
* Model compilation disabled
* GPU: single RTX 3090 24 GB

Relevant model settings:

```json
{
  "quantize": true,
  "qtype": "qfloat8",
  "quantize_te": true,
  "qtype_te": "qfloat8",
  "arch": "krea2",
  "low_vram": true,
  "compile": false,
  "layer_offloading": false
}
```

Dataset configuration tested:

```json
{
  "resolution": [
    1024,
    1280
  ],
  "cache_latents_to_disk": true,
  "num_frames": 1
}
```

I also tested lower resolutions that previously worked easily, but the new version still reaches the VRAM limit.

## Actual behavior

The model, transformer, text encoder and VAE load successfully. Dataset latents and text embeddings are cached successfully.

However, when training starts, almost the entire 24 GB VRAM is already occupied.

Example allocator warning:

```text
CUDACachingAllocator.cpp:3933] memory allocation failed with OOM on device 0
while trying to allocate 2097152 bytes
(free: 128188416, total: 25292177408)
```

A later allocation also fails:

```text
memory allocation failed with OOM on device 0
while trying to allocate 492830720 bytes
(free: 465829888, total: 25292177408)
```

In another run:

```text
memory allocation failed with OOM on device 0
while trying to allocate 492830720 bytes
(free: 471924736, total: 25292177408)
```

Training either stops or becomes unstable within the first few steps.

Example:

```text
emma: 0%| | 0/3800
emma: 0%| | 0/3800 [00:19<?, ?it/s, loss: 3.643e-02]
emma: 0%| | 1/3800 [00:36<38:54:21, 36.87s/it]
emma: 0%| | 2/3800 [00:46<24:30:56, 23.24s/it]
CUDACachingAllocator.cpp:3933] memory allocation failed with OOM
```

## Expected behavior

The training job should fit into 24 GB VRAM, as it did before the dependency update.

Before the update, the same GPU, model, LoRA rank, quantization settings and similar or higher dataset resolutions could train without OOM.

## Steps to reproduce

1. Use an RTX 3090 24 GB.
2. Install the current ai-toolkit requirements with PyTorch 2.13.0.
3. Load Krea 2 Raw using `qfloat8` quantization for both transformer and text encoder.
4. Enable gradient checkpointing.
5. Use batch size 1 and LoRA rank 32.
6. Cache latents and text embeddings.
7. Start training with a dataset resolution containing 1024 and/or 1280.
8. Observe that VRAM is almost completely full at the first training step.
9. Observe CUDA allocator OOM warnings and eventual training failure.

## Troubleshooting already attempted

* Full system reboot
* Starting from a clean GPU state
* `low_vram: true`
* `low_vram: false`
* Reduced training resolution
* Skipping the first sample
* Disabling compilation
* Caching latents to disk
* Caching text embeddings
* Unloading the text encoder after caching

None of these restored the previous VRAM behavior.

I also tried rolling back the PyTorch stack. The current ai-toolkit imports `torchaudio` globally, so a complete compatible Torch/Torchvision/Torchaudio version set is required. This makes partial rollback more difficult.

## Additional observation

The dataset is processed multiple times when several resolutions are listed, but the critical OOM occurs at the start of the actual forward/backward training steps, after the text encoder has already been unloaded.

The logs suggest that this is not simply leftover VRAM from another process. The new stack appears to use several hundred megabytes more VRAM or retain allocations that were previously released.

## Questions

* Is PyTorch 2.13 currently known to increase VRAM usage for Krea 2 training?
* Has the Krea 2 loading, quantization or gradient-checkpointing code changed recently?
* Is `low_vram` still expected to work without enabling `layer_offloading`?
* Could tensors from the VAE, quantization stage or cached embedding stage remain allocated when training begins?
* Would it be possible to keep PyTorch 2.12 compatibility or pin the requirements until this regression is resolved?

I can provide the complete training logs and configuration if needed.

[ai-toolkit_oom_log_1.txt](https://github.com/user-attachments/files/30648821/ai-toolkit_oom_log_1.txt)
[ai-toolkit_oom_log_0.txt](https://github.com/user-attachments/files/30648822/ai-toolkit_oom_log_0.txt)

**Comment by Tholue** [comment]
Update / confirmation:

After rolling back the PyTorch stack to:

- torch 2.11.0+cu130
- torchvision 0.26.0+cu130
- torchaudio 2.11.0+cu130
- triton 3.6.0

the same Krea 2 training job works again without OOM, including resolutions up to 1328, with low_vram disabled.

This strongly suggests that the VRAM regression is caused by the new PyTorch 2.13 / Triton 3.7 dependency stack rather than the dataset or training configuration.

**Comment by josh5671** [comment]
do you mind to share guide how to rollback as i have same issue? where i trained on 27jul, just now updated and it just oom. because i only know how to use the tool not the git/coding

**Comment by Tholue** [comment]
for my env it was like this:

uv pip uninstall --python .venv/bin/python torch torchvision torchaudio
uv pip install --python .venv/bin/python torch==2.11.0  torchvision==0.26.0 \ torchaudio==2.11.0 \ --index-url https://download.pytorch.org/whl/cu130

additionally this might help:
uv pip uninstall --python .venv/bin/python sageattention 

rm -rf .venv/lib/python3.12/site-packages/sageattention*

**Comment by josh5671** [comment]
it got a venv? i did 'figure out' something similar pip to install specific version but i stuck when i looking for venv to uninstall the broken torch
as i run on window thing might be different.....

**Comment by Tholue** [comment]
Yeah, somehow you have set up your Python environment to run ai-toolkit - also for Comfy. Asking any AI might help.

**Comment by josh5671** [comment]
With some reason, the installer i use didnt set venv on aitoolkit directory....
I think i might switch to other trainer as i have zero coding knowledge.


But still, appreciate your explanation

**Comment by josh5671** [comment]
i did manage to run with the torch and other 2~3 component version as you mention but Cuda12.8
problem persist even training krea 2 on 768. 
result as on my side. aitoolkit didnt OOM  even half hour or so without move a single step,
